hs_collection_resource/views.py

```diff
@@ -0,0 +1,105 @@
+            # the requesting user should at least have metadata view permission for each of the
+            # new resources to be added to the collection
+            for updated_contained_res_id in updated_contained_res_id_list:
+                if not collection_res_obj.resources.filter(short_id=updated_contained_res_id).exists():
+                    res_to_add, _, _ = authorize(request, updated_contained_res_id,
+                                                 needed_permission=ACTION_TO_AUTHORIZE.VIEW_METADATA)
+
+                    # for now we are not allowing a collection resource to be added to another collection resource
+                    if res_to_add.resource_type == "CollectionResource":
+                        raise Exception("Resource {0} is a collection resource which can't be added.".format(shortkey))
+
+            # remove all resources from the collection
+            collection_res_obj.resources.clear()
+
+            # add resources to the collection
+            for updated_contained_res_id in updated_contained_res_id_list:
+                updated_contained_res_obj = get_resource_by_shortkey(updated_contained_res_id)
+                collection_res_obj.resources.add(updated_contained_res_obj)
+
+            if collection_res_obj.can_be_public_or_discoverable:
+                metadata_status = "Sufficient to make public"
+
+            resource_modified(collection_res_obj, user)
+
+    except Exception as ex:
+        err_msg = "update_collection: {0} ; username: {1}; collection_id: {2} ."
+        logger.error(err_msg.format(ex.message,
+                     request.user.username if request.user.is_authenticated() else "anonymous",
+                     shortkey))
+        status = "error"
+        msg = ex.message
+    finally:
+        ajax_response_data = {'status': status, 'msg': msg, 'metadata_status': metadata_status}
+        return JsonResponse(ajax_response_data)
+
+
+def update_collection_for_deleted_resources(request, shortkey, *args, **kwargs):
+    """
+    If there are any tracked deleted resource objects for a collection resource
+    (identified by shortkey), those are deleted and resource bag is regenerated
+    for the collection resource to avoid the possibility of broken links in resource map
+    as a result of collection referenced resource being deleted by resource owner.
+    """
+
+    ajax_response_data = {'status': "success"}
+    try:
+        collection_res, is_authorized, user = authorize(request, shortkey,
+                                                        needed_permission=ACTION_TO_AUTHORIZE.EDIT_RESOURCE)
+
+        if collection_res.resource_type != "CollectionResource":
+            raise Exception("Resource {0} is not a collection resource.".format(shortkey))
+
+        resource_modified(collection_res, user)
+        # remove all logged deleted resources for the collection
+        collection_res.deleted_resources.all().delete()
```

Comment:
When a resource that belongs to a collection is deleted, we are not updating the resourcemap xml document for that collection resource. The reason we are not updating the collection resouremap, is to avoid the collection getting updated by a user (this the user who deleted the resource) who may not have any access permission for the collection. Not updating the collection resourcemap upon resource getting deleted, leaves broken links in the resource map document. By tracking the deleted resources we are able to inform the collection owners about resources being deleted and give them an opportunity to update the resourcemap.